### PR TITLE
Efficient yield in batches

### PR DESCRIPTION
**Pull request by montokapro**
### Summary

This PR intends to improve the efficiency of yielded relations from the `in_batches` method. Yielded relations now produce sql that more closely align with the original iteration pattern. Instead of defining an `IN` clause for all primary ids, a finish clause is added that bounds the last batch. `IN` queries with large batch sizes can be significantly less efficient. By replicating the initial iteration pattern closely, we can ensure the yielded relation is no less efficient than the original iteration pass. 

The original motivation for the `IN` query is [here](https://github.com/rails/rails/pull/20899#issuecomment-122541749). This PR also satisfies the original motivation.

The yielded relations are still subject to race conditions, though the behavior changes very slightly. The original implementation would guarantee that _the number of records returned by a yielded relation would be less or equal to the batch size_. This implementation does not guarantee this. It does, however, ensure that _all records within that batch interval are returned_, meaning that this could be more useful for dividing up work.

As a matter of potential impact, because the records are loaded into the yielded relation, it is unlikely for most callers to experience any change in behavior. To experience the effect of this change (beyond the reduced memory pressure of a simpler sql representation), one would need to further alter the relation to "unload" the existing records.

No new tests exist because no guarantees are expressed in documentation for which behavior changes. Beyond performance differences, behavior only changes under race conditions, which is expressed in the rdocs by `By its nature, batch processing is subject to race conditions if other processes are modifying the database.`

### Example

```
Post.in_batches(of: 2) Post.in_batches(of: 2) do |relation|
  puts relation.to_sql
end
```

#### Before
```
SELECT "posts".* FROM "posts" WHERE "posts"."id" IN (1, 2)
SELECT "posts".* FROM "posts" WHERE "posts"."id" IN (3, 4)
SELECT "posts".* FROM "posts" WHERE "posts"."id" IN (5, 6)
SELECT "posts".* FROM "posts" WHERE "posts"."id" IN (7, 8)
SELECT "posts".* FROM "posts" WHERE "posts"."id" IN (9, 10)
SELECT "posts".* FROM "posts" WHERE "posts"."id" = 11
```

#### After
```
SELECT "posts".* FROM "posts" WHERE "posts"."id" <= 2 ORDER BY "posts"."id" ASC
SELECT "posts".* FROM "posts" WHERE "posts"."id" > 2 AND "posts"."id" <= 4 ORDER BY "posts"."id" ASC
SELECT "posts".* FROM "posts" WHERE "posts"."id" > 4 AND "posts"."id" <= 6 ORDER BY "posts"."id" ASC
SELECT "posts".* FROM "posts" WHERE "posts"."id" > 6 AND "posts"."id" <= 8 ORDER BY "posts"."id" ASC
SELECT "posts".* FROM "posts" WHERE "posts"."id" > 8 AND "posts"."id" <= 10 ORDER BY "posts"."id" ASC
SELECT "posts".* FROM "posts" WHERE "posts"."id" > 10 AND "posts"."id" <= 11 ORDER BY "posts"."id" ASC
```

### Possible Variants

It would be possible to ensure the last yielded iteration does not have a finish (inclusive comparison) clause if no limit was provided. This mean that the final batch could become significantly larger than other batches with an append-only insert pattern under heavy write pressure, but it would guarantee that _the sum of all records from the yielded relations under a transaction would be equivalent to the records returned by the original relation under a transaction_. If this change is desired, I can implement it.

Alternately, if we would like to retain the guarantee of the original implementation that _the number of records returned by a yielded relation would be less or equal to the batch size_, then we can add a limit clause to each yielded relation.

### Other Information

This code change lays the ground work for two subsequent PRs:
https://github.com/montokapro/rails/pull/2 - Non-primary keys in batches
https://github.com/montokapro/rails/pull/1 - Composite keys in batches

The PR is meant to stand alone, regardless of whether the downstream PRs are accepted.

### CODE REVIEW

**Comment by brenogazzola** [comment]
Wouldn't this break if we add a `where` clause that made ids not continuous?

```ruby
author = Author.last

Post.where(author: author).in_batches(of: 2) do |relation|
  puts relation.to_sql
end

# SELECT "posts".* FROM "posts" WHERE "posts"."id" IN (1, 4)
# SELECT "posts".* FROM "posts" WHERE "posts"."id" IN (10, 23)
# SELECT "posts".* FROM "posts" WHERE "posts"."id" IN (100, 230)
```

**Comment by montokapro** [comment]
> Wouldn't this break if we add a `where` clause that made ids not continuous?
> 
> ```ruby
> Post.where(author: Current.user).in_batches(of: 2) do |relation|
>   puts relation.to_sql
> end
> 
> # SELECT "posts".* FROM "posts" WHERE "posts"."author_id" = 1 AND "posts"."id" IN (1, 5)
> # SELECT "posts".* FROM "posts" WHERE "posts"."author_id" = 1 AND "posts"."id" IN (13, 43)
> # SELECT "posts".* FROM "posts" WHERE "posts"."author_id" = 1 AND "posts"."id" IN (101, 600)
> ```

Good question! In this implementation, the last item returned becomes the start for the next batch. So the equivalent queries performed with this implementation would be:
```
# SELECT "posts".* FROM "posts" WHERE "posts"."author_id" = 1 AND "posts"."id" <= 5 ORDER BY "posts"."id" ASC
# SELECT "posts".* FROM "posts" WHERE "posts"."author_id" = 1 AND "posts"."id" > 5 AND "posts"."id" <= 43 ORDER BY "posts"."id" ASC
# SELECT "posts".* FROM "posts" WHERE "posts"."author_id" = 1 AND "posts"."id" > 43 AND "posts"."id" <= 600 ORDER BY "posts"."id" ASC
```

Under the possible variants section, I claim that we lose a guarantee provided from the original implementation with regards to non-continuous inputs. For example, if your data has a UUID as primary key, a batch could increase in size as new items to the database. To prevent this, we would need to add a limit to each yielded relation. However, this could leave gaps of unprocessed records.

In the original implementation, a similar problem could occur if a batch decreased in size because a record was deleted but still existed in the `IN` query - this could theoretically break an implementation that required an exact batch size.

**Comment by brenogazzola** [comment]
Oh yes, my bad. I was so focused on the non-continous id part I forgot that there would be an `author_id = 1` to prevent other entries from being returned.

**Comment by simi** [comment]
Is this similar to https://gitlab.com/gitlab-org/gitlab/-/blob/master/app/models/concerns/each_batch.rb?

**Comment by montokapro** [comment]
> Is this similar to https://gitlab.com/gitlab-org/gitlab/-/blob/master/app/models/concerns/each_batch.rb?

Yup! This looks like the same concept. The linked subsequent PRs add the ability to traverse a separate index, which I think has similarities to the `order-hint`.

I have a monkey-patched version of this running in production (with some additional enhancements alluded to by the subsequent PRs) that batch deletes millions of records every night.

File: activerecord/lib/active_record/relation/batches.rb [hunk 1]
```diff
@@ -221,27 +221,39 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
...
 
+      offset = nil
       loop do
-        if load
-          records = batch_relation.records
-          ids = records.map(&:id)
-          yielded_relation = where(primary_key => ids)
-          yielded_relation.load_records(records)
+        batch_relation = limit_relation
+        if offset
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation/batches.rb:231]:
```suggestion

        if offset
```

File: activerecord/lib/active_record/relation/batches.rb [hunk 2]
```diff
@@ -221,27 +221,39 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
...
-          yielded_relation = where(primary_key => ids)
+          batch_relation.pluck(primary_key)
         end
 
         break if ids.empty?
 
-        primary_key_offset = ids.last
-        raise ArgumentError.new("Primary key not included in the custom select clause") unless primary_key_offset
+        yielded_relation = relation
+        if offset
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation/batches.rb:246]:
```suggestion

        if offset
```

File: activerecord/lib/active_record/relation/batches.rb [hunk 3]
```diff
@@ -221,27 +221,39 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
...
+          yielded_relation = apply_offset_limit(yielded_relation, offset, order)
+        elsif start
+          yielded_relation = apply_start_limit(yielded_relation, start, order)
+        end
+
+        offset = ids.last
+        raise ArgumentError.new("Primary key not included in the custom select clause") unless offset
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation/batches.rb:253]:
```suggestion

        raise ArgumentError.new("Primary key not included in the custom select clause") unless offset
```

File: activerecord/lib/active_record/relation/batches.rb [hunk 4]
```diff
@@ -221,27 +221,39 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
...
 
+        yielded_relation = apply_finish_limit(yielded_relation, offset, order)
+        yielded_relation.load_records(batch_relation.records) if load
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation/batches.rb:256]:
```suggestion
        yielded_relation.load_records(batch_relation.records) if load

```

**Review by rafaelfranca — APPROVED** [review]
This changes make sense to me and it is indeed more performant in InnoDB.

@fxn what do you think?

**Comment by fxn** [comment]
@rafaelfranca on a first read, the idea sounds good to me. I have only two remarks:

1. I personally find the `limit` suffix in the name of the methods `apply_(offset|start|finish)_limit` a bit confusing. In isolation it makes sense, and their implementation is clear. However, in that context the name by itself may trigger they are about `LIMIT` SQL clauses. You have to go to the implementation to understand. I'd suggest something more specific like `add_offset_condition`, `add_offset_inequality`, or something like that.
2. If we are not loading the batch, can we do better than `ids = batch_relation.pluck(primary_key)`? Since we ask then `ids.empty?`, an array of IDs is fetched. However, now we do not need them all, we only need `MAX` or `MIN` depending on `order`.

**Comment by montokapro** [comment]
@rafaelfranca, @fxn - Thank you greatly for the thoughtful feedback and detailed consideration! Once I get a solid block of time, I will implement these suggestions. I appreciate you taking the time to validate the work of a new contributor :slightly_smiling_face: 

Per @rafaelfranca's second suggestion, I'd like to take a stab at this though I am initially sceptical this will be a simple change - rather I think it could lead to significant code duplication or complexity. I suspect this approach would require a subquery, since the database with need to scan the full batch size with a sql limit to find the batch end, and then extract the min/max from that query. I suspect this will reduce the amount of wire transfer time/bandwidth but not reduce the seek disk time. This might not be worth the implementation complexity... an analysis is warranted.

**Comment by montokapro** [comment]
@rafaelfranca @fxn - this is ready for re-review. (Apologizes if this ping is a burden or unwarranted - still learning conventions for contributing)

After this is merged, I have a follow-up PR with a straightforward bug fix for batching ordering I will submit for your (or anyone else's) review.

**Pull request closed without merging**